File: DataStructure_HandsOn/Java/Program9.java
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//	Program9:
//		A program which return second maximum element from singly linear linked list.
//
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import java.util.*;

class SinglyLL
{
	static int SecMaximum(LinkedList obj)
	{
		int iMax1 = 0,iMax2 = 0;

		Iterator <Integer>iObj = obj.iterator();
		Iterator <Integer>iObj1 = obj.iterator();
		while(iObj.hasNext())
		{
			int iVal = iObj.next();
			if(iMax1 < iVal)
			{
				iMax1 = iVal;
			}

		}
		
		while(iObj1.hasNext())
		{
			int iVal = iObj1.next();
			if(iMax2 < iVal && iMax1!=iVal)
			{
				iMax2 = iVal;
			}
		}	

		return iMax2;
	}

}

class Program9
{
	public static void main(String[] args)
	{
		LinkedList <Integer>lObj = new LinkedList<Integer>();
		int arr[] = new int[]{11,28,17,41,6,89};

		for(int i=0;i<6;i++)
		{
			lObj.add(arr[i]);
		}

		Iterator <Integer>iObj = lObj.iterator();
		System.out.print("Input Linked List: [");
		while(iObj.hasNext())
		{
			System.out.print(iObj.next()+", ");
		}

		System.out.println("]");
		
		int iRet = SinglyLL.SecMaximum(lObj);		
		System.out.println("Second Highest is: "+iRet);
	}
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//		Output:
//			Input Linked List: [11, 28, 17, 41, 6, 89, ]
//			Second Highest is: 41
//
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

File: DataStructure_HandsOn/C++/program9.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//	Program4:
//		A program which return second maximum element from singly linear linked list.
//
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include<iostream>
using namespace std;

typedef struct node
{
	int data;
	struct node* next;

}NODE,*PNODE;

class SinglyLL
{
	private:
		PNODE first;
		int no;
	
	public:
		SinglyLL(PNODE first=NULL,int no=0)
		{
		
		}

		void InsertFirst(int no)
		{
			PNODE newn = new NODE;

			newn->data = no;
			newn->next = NULL;

			if(first == NULL)
			{
				first = newn;
			}
			else
			{
				newn->next = first;
				first = newn;
			}
		}	

		
		int SecMaximum()
		{
			PNODE temp1 = first;
			PNODE temp2 = first;
			int iMax1=0,iMax2=0;

			while(temp1 != NULL)
			{
				if(iMax1 < temp1->data)
				{
					iMax1 = temp1->data;
				}

				temp1 = temp1->next;
			}

			while(temp2 != NULL)
			{
				if(iMax1 != temp2->data  && iMax2 < temp2->data)
				{
					iMax2 = temp2->data;
				}

				temp2 = temp2->next;
			}

			return iMax2;
		}	
	
	

		void Display()
		{
			PNODE temp = first;

			cout<<"Input Linked List: ";
			while(temp!=NULL)
			{
				cout<<"|"<<temp->data<<"|->";
				temp=temp->next;
			}
			cout<<"NULL\n";
		}
};


int main()
{
	int i = 0,iRet = 0;
	int arr[4] = {240,320,230,110};

	SinglyLL *sObj = new SinglyLL();
	

	for(i=0;i<4;i++)
	{
		sObj->InsertFirst(arr[i]);
	}	
	sObj->Display();
	
	iRet = sObj->SecMaximum();

	cout<<"Second Maximum is "<<iRet<<"\n";

	return 0;
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//	Output:
//		Input Linked List: |110|->|230|->|320|->|240|->NULL
//		Second Maximum is 240
//
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
